src/drivers/sifive,fe310-g000,pll.c

```diff
@@ -15,6 +18,156 @@
+    {20,  1,  6000000,  9600000, 0, 39, 2, -1}, 
+    {20,  1,  9600000, 12000000, 0, 19, 1, -1}, 
+    {20,  1, 12000000, 19200000, 1, 39, 1, -1}, 
+    {22,  1,  6000000,  8727272, 0, 43, 2, -1}, 
+    {22,  1,  8727273, 12000000, 0, 21, 1, -1}, 
+    {22,  1, 12000000, 17454545, 1, 43, 1, -1}, 
+    {24,  1,  6000000,  8000000, 0, 47, 2, -1}, 
+    {24,  1,  8000000, 12000000, 0, 23, 1, -1}, 
+    {24,  1, 12000000, 16000000, 1, 47, 1, -1}, 
+    {26,  1,  6000000,  7384615, 0, 51, 2, -1}, 
+    {26,  1,  7384616, 12000000, 0, 25, 1, -1}, 
+    {26,  1, 12000000, 14768230, 1, 51, 1, -1}, 
+    {28,  1,  6000000,  6857142, 0, 55, 2, -1}, 
+    {28,  1,  6857143, 12000000, 0, 27, 1, -1}, 
+    {28,  1, 12000000, 13714285, 1, 55, 1, -1}, 
+    {30,  1,  6000000,  6400000, 0, 59, 2, -1}, 
+    {30,  1,  6400000, 12000000, 0, 29, 1, -1}, 
+    {30,  1, 12000000, 12800000, 1, 59, 1, -1}, 
+    {32,  1,  6000000, 12000000, 0, 31, 1, -1}
+};
+
+#define PLL_CONFIG_NOT_VALID -1
+
+/* Given the rate of the PLL input frequency and a PLL configuration, what
+ * will the resulting PLL output frequency be?
+ * Arguments:
+ *  - pll_input_rate the PLL input frequency in hertz
+ *  - config the PLL configuration
+ * Returns:
+ *  - PLL_CONFIG_NOT_VALID if the configuration is not valid for the input frequency
+ *  - the output frequency, in hertz */
+static long get_pll_config_freq(long pll_input_rate, struct pll_config_t *config)
+{
+    if(pll_input_rate < config->min_input_rate || pll_input_rate > config->max_input_rate)
+        return PLL_CONFIG_NOT_VALID;
+
+    return pll_input_rate * config->multiplier / config->divisor;
+}
+
+#ifdef __MEE_DT_SIFIVE_FE310_G000_PLL_HANDLE
+
+static void mee_sifive_fe310_g000_pll_init(void) __attribute__((constructor));
+static void mee_sifive_fe310_g000_pll_init(void) {
+    __mee_driver_sifive_fe310_g000_pll_init(__MEE_DT_SIFIVE_FE310_G000_PLL_HANDLE);
+}
+
+#endif /* __MEE_DT_SIFIVE_FE310_G000__PLL_HANDLE */
+
+void __mee_driver_sifive_fe310_g000_pll_init(struct __mee_driver_sifive_fe310_g000_pll *pll) {
+    mee_io_u32 *pllcfg = (mee_io_u32 *) (pll->config_base->base + pll->config_offset);
+
+    /* If the PLL clock has had a pre_rate_change_callback configured, call it */
+    if(pll->clock.pre_rate_change_callback != NULL)
+        pll->clock.pre_rate_change_callback(pll->clock.pre_rate_change_callback_priv);
+
+    /* If we're running off of the PLL, switch off before we start configuring it*/
+    if((MEE_ACCESS_ONCE(pllcfg) & PLL_SEL) == 0)
+        MEE_ACCESS_ONCE(pllcfg) &= ~(PLL_SEL);
+
+    /* Make sure we're running off of the external oscillator for stability */
+    if(pll->pllref != NULL)
+        MEE_ACCESS_ONCE(pllcfg) |= PLL_REFSEL;
+
+    /* Configure the PLL to run at the requested init frequency.
+     * Using the vtable instead of the user API because we want to control
+     * when the callbacks occur. */
+    pll->vtable->clock.set_rate_hz(&(pll->clock), pll->init_rate);
```

Comment:
You should allow `pll->init_rate` to be 0, in case a user just wants to use whatever is baked into the hardware.